### PR TITLE
fix: Check if every Issue have project assigned.

### LINKED ISSUE #157: Ensure that every issue is assigned to a project
Ensure that every issue is assigned to a project. This should be done with two things:

1. Oppiabot comments on new issues that they need to be assigned to some project
2. Periodic job will comment on issues that are not yet assigned to some project

### PR DESCRIPTION
**Pull request by Naman-1234**
## Explanation

Fixes #157 
## Checklist
- [x] I have successfully deployed my own instance of Oppiabot.
  - You can find instructions for doing this [here](https://github.com/oppia/oppiabot/wiki/Deploying-your-own-instance-of-the-oppiabot).
- [x] I have manually tested all the changes made in this PR following the [manual tests matrix](https://github.com/oppia/oppiabot/wiki/Manual-Tests-Matrix).

### CODE REVIEW

**Comment by Naman-1234** [comment]
@vojtechjelinek  and @jameesjohn  PTAL Sir, Could you please review this Pr. I have also tested it manually on one of My repos Named Eclec, Which I removed while Performing Pr. Thanks.
Please refer to the following video:

https://user-images.githubusercontent.com/63748249/113019597-6aa5a400-919f-11eb-9e06-4e7d3e366815.mp4

**Comment by jameesjohn** [comment]
Can you assign the Issue to a project while creating the issue and make a video of that?

File: constants.js
Comment (jameesjohn) [inline constants.js:134]: Since this is run when an issue is opened, should this be `ensureNewIssuesHaveProjects`?

**Review by jameesjohn — CHANGES_REQUESTED** [review]

File: constants.js
Comment (Naman-1234) [reply constants.js:134]: Sir Actually I have a doubt to clarify here, Current the function  which I am using(Was written already),WIll comment on all the issues which do not have project assigned, Do you want that this should be restricted to only newly opened issues. Please let me know, So I can proceed in this PR.

**Comment by Naman-1234** [comment]
> Can you assign the Issue to a project while creating the issue and make a video of that?

Yes Sir I have made it, Pls have a look here:

#### In the following video, First a user does not add a project, Due to which oppiabot comments to remind the user of adding a project, then in next one user adds a project, So It does not comment.
https://user-images.githubusercontent.com/63748249/113140006-7350b580-9245-11eb-8718-65248495b6ca.mp4

File: constants.js
Comment (vojtechjelinek) [reply constants.js:134]: Yep, I think we should only check the issue that triggered the event.

**Comment by Naman-1234** [comment]
@jameesjohn  and @vojtechjelinek  Sir Please review this PR. I think the issue have been targeted. Please find the attached video.Thanks.

##### In the video First I made a issue and didn't attach any project, So Oppiabot commented but in the next one I attached the project and hence Oppiabot didn't comment.

https://user-images.githubusercontent.com/63748249/113216990-cd2f9a80-929a-11eb-8ed7-5503f8a257b1.mp4

File: lib/periodicChecks.js
Comment (vojtechjelinek) [inline lib/periodicChecks.js:184]: Why this sleep?

File: lib/periodicChecks.js
Comment (vojtechjelinek) [inline lib/periodicChecks.js:193]: This comment is not needed, the purpose of `doesIssueHaveProject` is pretty clear.

File: lib/periodicChecks.js
Comment (vojtechjelinek) [inline lib/periodicChecks.js:202]:
```suggestion
          'any project. Can you please assign it to some project? Thanks!',
```

File: lib/periodicChecks.js
Comment (vojtechjelinek) [inline lib/periodicChecks.js:190]: Isn't there a different way to do this? Can't we get the project to which the issues is assigned directly from the issue object?

**Review by vojtechjelinek — CHANGES_REQUESTED** [review]
Thanks! Left a few comments.

File: lib/periodicChecks.js
Comment (Naman-1234) [reply lib/periodicChecks.js:184]: This sleep is necessary here, Because this is providing time to user to add project, Otherwise if this is not used then as soon as Issue gets created Oppiabot will comment to add project and will not provide user time to add project before.

File: lib/periodicChecks.js
Comment (Naman-1234) [reply lib/periodicChecks.js:193]: Ok Sir, Removing it in latest commit.

File: lib/periodicChecks.js
Comment (Naman-1234) [reply lib/periodicChecks.js:202]: Updating it in the latest commit Sir.

File: lib/periodicChecks.js
Comment (Naman-1234) [reply lib/periodicChecks.js:190]: No Sir, There is no such way As issue object does not have any field related to project.

File: lib/periodicChecks.js
Comment (vojtechjelinek) [reply lib/periodicChecks.js:184]: Hmm, okay makes sense, this is 20 second right?

Comment (vojtechjelinek) [reply lib/periodicChecks.js:184]: Probably add a comment in the code about this.

File: lib/periodicChecks.js
Comment (Naman-1234) [reply lib/periodicChecks.js]: Sure Sir, Adding them right now.

File: lib/periodicChecks.js
Comment (Naman-1234) [reply lib/periodicChecks.js:184]: Sure Sir, Adding them right now.

Comment (Naman-1234) [reply lib/periodicChecks.js:184]: Is there any other change required Sir, Or Should I perform these changes and perform PR?

Comment (Naman-1234) [reply lib/periodicChecks.js:184]: > Hmm, okay makes sense, this is 20 second right?

Yes sir.

**Comment by Naman-1234** [comment]
@vojtechjelinek  and @jameesjohn  PTAL Sir. Reviews have been addressed. Thanks.

**Comment by Naman-1234** [comment]
@vojtechjelinek  and @jameesjohn  PTAL Sir, All the changes requested have been addressed Sir. Please review this. Thanks.

File: constants.js
Comment (vojtechjelinek) [inline constants.js:103]: I don't understand this, I thought that in this object we had a mapping between events and checks, why you map `ensureNewIssuesHaveProjects` to `ensureNewIssuesHaveProjects`?

File: constants.js
Comment (vojtechjelinek) [inline constants.js:75]: Unneeded change.

File: lib/periodicChecks.js
Comment (vojtechjelinek) [inline lib/periodicChecks.js:180]:
```suggestion
 * 1. Check if the new issue has a project assigned.
 * 2. If not then ping the Core Maintainers.
```

File: lib/periodicChecks.js
Comment (vojtechjelinek) [inline lib/periodicChecks.js:186]:
```suggestion
  // This sleep function is used to provide the user some time to add project 
  // after making the issue.
```

**Review by vojtechjelinek — CHANGES_REQUESTED** [review]
Thanks! Left a few comments.

File: constants.js
Comment (Naman-1234) [reply constants.js:75]: Sorry for that, It was a mistake,Correcting it in the latest commit.

File: lib/periodicChecks.js
Comment (Naman-1234) [reply lib/periodicChecks.js:180]: Changing it in the latest Commit.

File: lib/periodicChecks.js
Comment (Naman-1234) [reply lib/periodicChecks.js:186]: Changing it in the latest Commit.

File: constants.js
Comment (Naman-1234) [reply constants.js:103]: Actually in this I took names of both same, But will change it in the latest commit. Thanks for telling.

**Comment by Naman-1234** [comment]
@vojtechjelinek  and @jameesjohn PTAL Sir, Changes Requested have been Targeted sir, Please review this. Thanks.

**Review by jameesjohn — CHANGES_REQUESTED** [review]
Thanks, I left some comments. Also, tests have not been written for this functionality.

File: constants.js
Comment (Naman-1234) [reply constants.js:29]: You mean changing name to issueOpenedEvent?

Comment (Naman-1234) [reply constants.js]: Changed it in the latest commit Sir.

File: constants.js
Comment (vojtechjelinek) [inline constants.js:160]:
```suggestion
module.exports.ensureNewIssuesHaveProjectsCheck = (
  ensureNewIssuesHaveProjectsCheck);
```

File: index.js
Comment (vojtechjelinek) [inline index.js:161]:
```suggestion
            callable.push(
              periodicCheckModule.ensureNewIssuesHaveProjects(context));

```

**Review by vojtechjelinek — CHANGES_REQUESTED** [review]
THanks! A few last notes.

File: constants.js
Comment (Naman-1234) [reply constants.js:160]: Will change them in the latest commit Sir.Thanks.

File: index.js
Comment (Naman-1234) [reply index.js:161]: Will change them in the latest commit Sir.Thanks.

File: constants.js
Comment (vojtechjelinek) [inline constants.js:135]:
```suggestion
module.exports.issueOpenedEvent = issueOpenedEvent;
```

**Review by vojtechjelinek — CHANGES_REQUESTED** [review]
One last note.

File: constants.js
Comment (Naman-1234) [reply constants.js]: Changed it in the latest commit Sir.

**Review by vojtechjelinek — APPROVED** [review]
LGTM!

**Comment by Naman-1234** [comment]
@vojtechjelinek Sir I think I need to ask, Should I remove this Check every issue has a project from Periodic Checks(Was Already in the codebase). Since it is checking at the time of the new issue, There is no need to check it periodically.

**Comment by vojtechjelinek** [comment]
> @vojtechjelinek Sir I think I need to ask, Should I remove this Check every issue has a project from Periodic Checks(Was Already in the codebase). Since it is checking at the time of the new issue, There is no need to check it periodically.

Keep it, it can happen that someone removes the project and then it is worthwhile to notify.

**Review by jameesjohn — APPROVED** [review]
LGTM

**Comment by jameesjohn** [comment]
Will merge after #254 gets in.

**Review by jameesjohn — CHANGES_REQUESTED** [review]
Hi @Naman-1234, tests have not been written for this feature?

**Comment by Naman-1234** [comment]
@jameesjohn  Sir Actually the tests are already written, Were written few days, Didn't know why they were not showing but I pulled the changes and now they are coming perfectly.
Could you please review this PR now. Thanks.

**Comment by Naman-1234** [comment]
@jameesjohn  Sir Could you please take a look at this PR(Tests successfully written) and tell me if any work is needed on this PR. This will help me in proceeding with this PR. Thanks.

**Comment by Naman-1234** [comment]
@jameesjohn Sir Sorry for the disturbance. Could you please take a look at this PR now. Now the tests are there and also passing(Don't know why they were not showing before, After pull and push they happened to show).
Please take a look and let me know if any change is required in them, This will help me in proceeding in this PR. Thanks.

**Comment by Naman-1234** [comment]
@vojtechjelinek Could you please assign james Sir on this one also. It will bring to his kind notice about this PR. Thanks.

File: spec/periodicChecksSpec.js
Comment (jameesjohn) [inline spec/periodicChecksSpec.js:873]: The payload for this event should be `issued.opened` since this feature works when an issue is opened. The default `payloadData` event and action si `schedule.repository`. https://github.com/oppia/oppiabot/blob/a9c06dc340450a5d40cd3aa42bebaec421650b7c/fixtures/periodicCheckPayload.json#L1-L4

**Review by jameesjohn — CHANGES_REQUESTED** [review]

File: spec/periodicChecksSpec.js
Comment (Naman-1234) [reply spec/periodicChecksSpec.js:873]: Sure Sir, Changing it right now. Please let me know if any other change is required. I will be happy to fix. Thanks.

**Comment by Naman-1234** [comment]
@jameesjohn Sir Changes have been targeted. Please Take a look. Thanks.

**Comment by Naman-1234** [comment]
@vojtechjelinek  and @jameesjohn This is not related to this issue. But I want to ask are there any periodic meetings. if yes I would like to be a part of them if you grant so. Hoping for an reply. Thanks.

**Comment by jameesjohn** [comment]
Hi @Naman-1234, yes we usually have weekly meetings. We actually have one tomorrow. Can I have your email address?

File: spec/periodicChecksSpec.js
Comment (jameesjohn) [reply spec/periodicChecksSpec.js:875]: Sorry I had a typo in my earlier message. It should be `issues.opened` not `issued`.

**Comment by Naman-1234** [comment]
@jameesjohn  Sure Sir My Email address is `namankalrabhiwani54@gmail.com` . Thanks for Responding Sir.

File: spec/periodicChecksSpec.js
Comment (Naman-1234) [reply spec/periodicChecksSpec.js:875]: Changing it right away Sir. Thanks.

**Comment by Naman-1234** [comment]
@jameesjohn  Changes have been targeted sir. Please take a look. Thanks.

**Comment by Naman-1234** [comment]
> @jameesjohn Sure Sir My Email address is `namankalrabhiwani54@gmail.com`. Thanks for Responding Sir.

@vojtechjelinek  and @jameesjohn sir it's just a health reminder in case you forgot. I have not received any mail till now and today is the meeting.Could you please take a look at this.Thanks.

**Comment by jameesjohn** [comment]
The invite has been sent.
Also, check your hangouts for invitations to join the oppiabot team group chat.

**Comment by Naman-1234** [comment]
> The invite has been sent.
> Also, check your hangouts for invitations to join the oppiabot team group chat.

I accepted it, Thank you so much, Sir. 
@jameesjohn Sir, Could you please also provide an answer to one other question. Why I am not able to assign because all other contributors are able to do so?  Does it require some permissions, If yes Could you please provide them. That way I will be able to assign also like other contributors. Looking forward for a reply. Thanks.

**Comment by Naman-1234** [comment]
@jameesjohn  All the work on this Pr is been done. Please take a look. Thanks.
Also, Sir As agreed in the meeting, Could you please promote me and provide me assigning privileges.

**Comment by Naman-1234** [comment]
@vojtechjelinek Please assign james sir on this.

File: spec/periodicChecksSpec.js
Comment (jameesjohn) [inline spec/periodicChecksSpec.js:887]: These tests look wrong.
Are these what we expect to happen?

**Review by jameesjohn — CHANGES_REQUESTED** [review]

File: spec/periodicChecksSpec.js
Comment (Naman-1234) [reply spec/periodicChecksSpec.js:887]: It looks right to me Because we have to first get all the project cards then filter out the ones required. Hence a test is required to check that all the project cards are fetched.
Looking forward to your reviews on this sir. Thanks.

**Comment by Naman-1234** [comment]
@jameesjohn  Sir Could you please let me know about my reply to your comment If it still seems wrong to you. then I would modify/remove those tests. Thanks.

**Comment by Naman-1234** [comment]
@jameesjohn  I have added the response in case issue is assigned to a project. In this case other than the tests already written in this Pull Request an additional behaviour is it should not ping core maintainers. Please let me know if any change is requested.Thanks.

File: spec/periodicChecksSpec.js
Comment (jameesjohn) [inline spec/periodicChecksSpec.js:875]: The `payloadData` should contain some information about the issue. See the `issue.assigned.json` payload.
If possible, use that as the payload and set the action to be opened.

Comment (jameesjohn) [reply spec/periodicChecksSpec.js:875]: Do this here and in the other test.

**Review by jameesjohn — CHANGES_REQUESTED** [review]

File: spec/periodicChecksSpec.js
Comment (Naman-1234) [reply spec/periodicChecksSpec.js:875]: Working on it. Thanks.

**Comment by Naman-1234** [comment]
@jameesjohn  Sir I replaced payload, But now tests are failing. I am not getting what is the reason behind that. Could you please help me out here a little. Thanks.

File: spec/periodicChecksSpec.js
Comment (jameesjohn) [reply spec/periodicChecksSpec.js:926]: This should make the tests pass in isolation, but for some reason, it doesn't pass when run with other tests.

**Review by jameesjohn — CHANGES_REQUESTED** [review]

File: spec/periodicChecksSpec.js
Comment (Naman-1234) [reply spec/periodicChecksSpec.js:926]: Performing the changes in the latest commit Sir. Thanks for guiding me in the right direction.

**Pull request closed without merging**
